x-pack/plugins/apm/public/utils/timepicker/index.js

```diff
@@ -49,55 +48,41 @@ export function initTimepicker(history, dispatch, callback) {
       });
 
       history.listen(() => {
-        updateRefreshRate(dispatch, timefilter);
+        updateRefreshRate(dispatch);
         globalState.fetch();
       });
-      timefilter.setTime = (from, to) => {
-        timefilter.time.from = moment(from).toISOString();
-        timefilter.time.to = moment(to).toISOString();
-        $scope.$apply();
-      };
       timefilter.enableTimeRangeSelector();
       timefilter.enableAutoRefreshSelector();
-      timefilter.init();
 
-      updateRefreshRate(dispatch, timefilter);
+      updateRefreshRate(dispatch);
 
-      timefilter.on('update', () => dispatch(getAction(timefilter)));
+      $scope.$listen(timefilter, 'timeUpdate', () =>
+        dispatch(updateTimePickerAction())
+      );
 
-      // hack to access timefilter outside Angular
-      globalTimefilter = timefilter;
+      registerTimefilterWithGlobalState(globalState);
```

Comment:
Instead of this, let's just call the method directly (again, less coupling to Angular):
```js
dispatch(updateTimePickerAction())
```

Reply:
`registerTimefilterWithGlobalState ` needs to be called. It sets the time and refresh interval from the URL and places a listener on globalState so if the URL gets updated - then the time and refresh interval get updated to reflect the URL.

Reply:
Ah, okay. 

Since timefilter automatically updates the url on change, shouldn't it also listen for url changes? It seems a little odd that it only works one way, and to use it the consumer has to setup a url listener manually. It would also give us a nice single-source-of-truth, where time is just derived off of the url - whereas now it seems time state is stored in both url, globalState and timefilter itself.

Probably doesn't fit in this PR though :)

Reply:
Timefilter does listen to URL changes. `registerTimefilterWithGlobalState ` sets up the listener. Its just that the current implementation is based on angular with the globalState service and subscribes to an event on globalState. Refactoring this relationship will need to be done later - progress over perfection